hoodie-utilities/src/main/java/com/uber/hoodie/utilities/deltastreamer/DeltaSync.java

```diff
@@ -0,0 +1,552 @@
+   * @param error
+   */
+  private void shutdownCompactor(boolean error) {
+    log.info("Delta Sync shutdown. Error ?" + error);
+    if (compactor != null) {
+      log.warn("Gracefully shutting down compactor");
+      compactor.shutdown(false);
+    }
+  }
+
+  /**
+   * Run one round of ingestion and return
+   * @throws Exception
+   */
+  public void syncOneRound() throws Exception {
+    HoodieDeltaStreamerMetrics metrics = new HoodieDeltaStreamerMetrics(getHoodieClientConfig(schemaProvider));
+    Timer.Context overallTimerContext = metrics.getOverallTimerContext();
+
+    // Refresh Timeline
+    refreshTimeline();
+
+    Pair<SchemaProvider, Pair<String, JavaRDD<HoodieRecord>>> srcRecordsWithCkpt = readFromSource();
+
+    if (null != srcRecordsWithCkpt) {
+      // this is the first input batch. If schemaProvider not set, use it and register Avro Schema and start
+      // compactor
+      if (null == schemaProvider) {
+        // Set the schemaProvider if not user-provided
+        this.schemaProvider = srcRecordsWithCkpt.getKey();
+        // Setup HoodieWriteClient and compaction now that we decided on schema
+        setupWriteClientAndStartCompaction();
+      }
+
+      writeToSink(srcRecordsWithCkpt.getRight().getRight(), srcRecordsWithCkpt.getRight().getLeft(), metrics,
+          overallTimerContext);
+    }
+
+    // Clear persistent RDDs
+    jssc.getPersistentRDDs().values().forEach(JavaRDD::unpersist);
+  }
+
+  /**
+   * Note that depending on configs and source-type, schemaProvider could either be eagerly or lazily created.
+   * SchemaProvider creation is a precursor to HoodieWriteClient and AsyncCompactor creation. This method takes
+   * care of this contraint.
+   */
+  private void setupWriteClientAndStartCompaction() {
+    log.info("Setting up Hoodie Write Client");
+    if ((null != schemaProvider) && (null == writeClient)) {
+      registerAvroSchemas(schemaProvider);
+      HoodieWriteConfig hoodieCfg = getHoodieClientConfig(schemaProvider);
+      writeClient = new HoodieWriteClient<>(jssc, hoodieCfg, true);
+      if (tableType.equals(HoodieTableType.MERGE_ON_READ)) {
+        compactor = new AsyncCompactor(jssc, writeClient);
+        // Enqueue existing pending compactions first
+        HoodieTableMetaClient meta = new HoodieTableMetaClient(
+            new Configuration(fs.getConf()), cfg.targetBasePath, true);
+        List<HoodieInstant> pending = CompactionUtils.getPendingCompactionInstantTimes(meta);
+        pending.stream().forEach(hoodieInstant -> compactor.enqueuePendingCompaction(hoodieInstant));
+        compactor.start((error) -> {
+          // Shutdown DeltaSync
+          shutdown(false);
+          return true;
+        });
+      }
+    }
+  }
+
+  /**
+   * Read from Upstream Source and apply transformation if needed
+   * @return
+   * @throws Exception
+   */
+  private Pair<SchemaProvider, Pair<String, JavaRDD<HoodieRecord>>> readFromSource() throws Exception {
```

Comment:
is this mostly code that was moved out of HoodieDeltaStreamer?

Reply:
Yes